File: solutions/topological-sort/630-course.py
"""
# 630. Course
# Difficulty: Medium
Given a problem that demonstrates key concepts in Topological Sort.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of topological sort concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply topological sort methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages topological sort principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1)

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

The approach uses topological sort techniques to solve this problem efficiently.

### Algorithm Steps:
1. Initialize necessary variables
2. Process input using topological sort method
3. Return the computed result

</details>
"""

class Solution:
    def scheduleCourse(self, courses: List[List[int]]) -> int:
        """
        Finds the maximum number of courses that can be taken given their durations
        and deadlines.

        Args:
            courses: List of [duration, deadline] pairs for each course

        Returns:
            Maximum number of courses that can be taken

        Time Complexity: O(n log n) where n is the number of courses
        Space Complexity: O(n) for storing courses in the heap
        """
        # Sort courses by deadline to process earlier deadlines first
        courses.sort(key=lambda x: x[1])
        
        # Max heap to store course durations we've taken
        taken = []
        current_time = 0
        
        for duration, deadline in courses:
            # If we can take this course without exceeding its deadline
            if current_time + duration <= deadline:
                current_time += duration
                # Push negative duration for max heap (Python has min heap by default)
                heapq.heappush(taken, -duration)
            # If we can't take it directly, check if we can swap with a longer course
            elif taken and -taken[0] > duration:
                # Remove the longest course we've taken so far
                longest_duration = -heapq.heappop(taken)
                # Add the time difference after swapping
                current_time = current_time - longest_duration + duration
                # Add the new course
                heapq.heappush(taken, -duration)
        
        # Return the number of courses taken
        return len(taken)

def test_solution():
    """
    Test cases for 630. Course.
    """
    solution = Solution()

    # Test case 1: Basic functionality
    # result = solution.solve([test_input])
    # expected = [expected_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Edge case
    # result = solution.solve([edge_case_input])
    # expected = [edge_case_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 630. Course")

File: docs/solutions/topological-sort/alternatives/630-course.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 * **Step 1:** [description]
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **Pointers meet:** Handle when left == right
 * - **Empty input:** Check for null or empty arrays
 * - **Single element:** One pointer scenario
 * - **All duplicates:** Pointer movement with same values
 * - **Boundary crossing:** Prevent left > right
 *
 * </details>
 */

/**
 * Main solution for Problem 630: Course Schedule III
 *
 * @param {number[][]} courses - Array of [duration, lastDay] pairs
 * @return {number} - Maximum number of courses that can be taken
 *
 * Time Complexity: O(n log n) for sorting and priority queue operations
 * Space Complexity: O(n) for the priority queue
 */
function solve(courses) {
    // Sort courses by their end day (deadline)
    courses.sort((a, b) => a[1] - b[1]);

    // Max heap to store durations (using negative values for max behavior)
    const maxHeap = [];
    let currentTime = 0;

    // Helper functions for max heap
    const heapPush = (val) => {
        maxHeap.push(val);
        let idx = maxHeap.length - 1;
        while (idx > 0) {
            const parent = Math.floor((idx - 1) / 2);
            if (maxHeap[parent] >= maxHeap[idx]) break;
            [maxHeap[parent], maxHeap[idx]] = [maxHeap[idx], maxHeap[parent]];
            idx = parent;
        }
    };

    const heapPop = () => {
        if (maxHeap.length === 0) return null;
        if (maxHeap.length === 1) return maxHeap.pop();

        const max = maxHeap[0];
        maxHeap[0] = maxHeap.pop();
        let idx = 0;

        while (true) {
            const left = 2 * idx + 1;
            const right = 2 * idx + 2;
            let largest = idx;

            if (left < maxHeap.length && maxHeap[left] > maxHeap[largest]) {
                largest = left;
            }
            if (right < maxHeap.length && maxHeap[right] > maxHeap[largest]) {
                largest = right;
            }
            if (largest === idx) break;

            [maxHeap[idx], maxHeap[largest]] = [maxHeap[largest], maxHeap[idx]];
            idx = largest;
        }

        return max;
    };

    // Greedy approach: take courses, replace with shorter ones if needed
    for (const [duration, lastDay] of courses) {
        currentTime += duration;
        heapPush(duration);

        // If we exceed the deadline, remove the longest course taken so far
        if (currentTime > lastDay) {
            currentTime -= heapPop();
        }
    }

    return maxHeap.length;
}

/**
 * Test cases for Problem 630: Course Schedule III
 */
function testSolution() {
    console.log('Testing 630. Course Schedule III');

    // Test case 1: Can take 3 courses
    const result1 = solve([[100, 200], [200, 1300], [1000, 1250], [2000, 3200]]);
    const expected1 = 3;
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);

    // Test case 2: Can only take 1 course
    const result2 = solve([[1, 2]]);
    const expected2 = 1;
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);

    // Test case 3: Can take 2 courses
    const result3 = solve([[3, 2], [4, 3]]);
    const expected3 = 0;
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);

    // Test case 4: Multiple courses with same deadline
    const result4 = solve([[5, 5], [4, 6], [2, 6]]);
    const expected4 = 2;
    console.assert(result4 === expected4, `Test 4 failed: expected ${expected4}, got ${result4}`);

    // Test case 5: All courses can be taken
    const result5 = solve([[1, 2], [2, 4], [3, 7]]);
    const expected5 = 3;
    console.assert(result5 === expected5, `Test 5 failed: expected ${expected5}, got ${result5}`);

    console.log('All test cases passed for 630. Course Schedule III!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 630. Course ===');
    console.log('Category: Topological Sort');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on topological sort concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */
